pkg/logs/client/tcp/destination.go

```diff
@@ -13,113 +13,128 @@ import (
+		isRetrying:          false,
 	}
 }
 
-// Send transforms a message into a frame and sends it to a remote server,
-// returns an error if the operation failed.
-func (d *Destination) Send(payload []byte) error {
-	if d.conn == nil {
-		var err error
-
-		// We work only if we have a started destination context
-		ctx := d.destinationsContext.Context()
-		if d.conn, err = d.connManager.NewConnection(ctx); err != nil {
-			// the connection manager is not meant to fail,
-			// this can happen only when the context is cancelled.
-			return err
+// Start reads from the input, transforms a message into a frame and sends it to a remote server,
+func (d *Destination) Start(input chan *message.Payload, output chan *message.Payload) (stopChan chan struct{}) {
+	stopChan = make(chan struct{})
+	go func() {
+		for payload := range input {
+			d.sendAndRetry(payload, output)
 		}
-		d.connCreationTime = time.Now()
-	}
-
-	metrics.BytesSent.Add(int64(len(payload)))
-	metrics.TlmBytesSent.Add(float64(len(payload)))
-	metrics.EncodedBytesSent.Add(int64(len(payload)))
-	metrics.TlmEncodedBytesSent.Add(float64(len(payload)))
+		d.Lock()
+		d.isRetrying = false
+		d.Unlock()
+		stopChan <- struct{}{}
+	}()
+	return stopChan
+}
 
-	content := d.prefixer.apply(payload)
-	frame, err := d.delimiter.delimit(content)
-	if err != nil {
-		// the delimiter can fail when the payload can not be framed correctly.
-		return err
-	}
+// GetIsRetrying returns true if the destination is retrying
+func (d *Destination) GetIsRetrying() bool {
+	d.Lock()
+	defer d.Unlock()
+	return d.isRetrying
+}
 
-	_, err = d.conn.Write(frame)
-	if err != nil {
-		d.connManager.CloseConnection(d.conn)
-		d.conn = nil
-		return client.NewRetryableError(err)
-	}
+func (d *Destination) sendAndRetry(payload *message.Payload, output chan *message.Payload) {
+	for {
+		if d.conn == nil {
+			var err error
+
+			// We work only if we have a started destination context
+			ctx := d.destinationsContext.Context()
+			if d.conn, err = d.connManager.NewConnection(ctx); err != nil {
+				// the connection manager is not meant to fail,
+				// this can happen only when the context is cancelled.
+				d.incrementErrors(true)
+				return
+			}
+			d.connCreationTime = time.Now()
+		}
 
-	if d.connManager.ShouldReset(d.connCreationTime) {
-		log.Debug("Resetting TCP connection")
-		d.connManager.CloseConnection(d.conn)
-		d.conn = nil
-	}
+		content := d.prefixer.apply(payload.Encoded)
+		frame, err := d.delimiter.delimit(content)
+		if err != nil {
+			// the delimiter can fail when the payload can not be framed correctly.
+			d.incrementErrors(true)
+			return
+		}
 
-	return nil
-}
+		_, err = d.conn.Write(frame)
+		if err != nil {
+			d.connManager.CloseConnection(d.conn)
+			d.conn = nil
+
+			if d.shouldRetry {
+				d.Lock()
+				d.isRetrying = true
+				d.Unlock()
+				d.incrementErrors(false)
+				// retry (will try to open a new connection)
+				continue
+			} else {
+				d.incrementErrors(true)
```

Comment:
No need for `else` - continue should end the logic flow for the other cases I think.